feat: Refactor bpftrace functionality to generic tracer

zqureshi opened the pull request:
This PR introduces the tracer interface as described in the [RFC](https://docs.google.com/document/d/1Ip4D_vc-aOyVP2EsvgF_sxgBNngTPrxPk8Ybw6j4_6g/edit) and #125 while maintaining full backwards compatibility.

The goal was to make the minimal amount of changes that would get us to the new world but leave more feature implementations to future PRs. While working on the code and after some discussions I also tweaked the interface to remove the `--target` flag and just have a single `--selector` flag.

Review guide:
- Review `TraceRunnerOptions` and verify new fields.
- Review `RunOptions` changes, focusing on the new fields that mirror tracerunner.
- I renamed image from `kubectl-trace-bpftrace` to `kubectl-trace-runner`.
- Review [Selector](https://github.com/iovisor/kubectl-trace/pull/136/files#diff-b57f265d7e29aabfcb36e32ada2984df867c3b7e81d1b95e26539b0da070ca12) and accessors. This represents a k8s label query but with accessors for each field.
- Go to [tracerunner](https://github.com/iovisor/kubectl-trace/pull/136/files#diff-43ad48182bb415af9c97abf39d10fd3e78a2755559f0d635da6dd3020a22be52R127) and review the `Validate()` and `Run()` methods.

  For a simple `kubectl trace -e '<bpf_script>' minikube` the generated runner command will be `trace-runner --tracer=bpftrace --selector=node=minikube --output=stdout --program=/programs/program.bt`

  When selecting a pod / container the generated selector will have `pod-uid=<uuid>,container=<name>` so review the `$container_pid` substitution that leverages [that](https://github.com/iovisor/kubectl-trace/pull/136/files#diff-43ad48182bb415af9c97abf39d10fd3e78a2755559f0d635da6dd3020a22be52R184).
- Now get to [run.go](https://github.com/iovisor/kubectl-trace/pull/136/files#diff-d9b5c5a6245814d60e99f92a71feb162ec636c09f5787321970e26ef80fb736bR239) and review the `Complete()` method changes and how [job.go](https://github.com/iovisor/kubectl-trace/pull/136/files#diff-6c50df8d5924b03f85013c687895c9f95d43d0f704a18ae6abbf752ea6d7c916R189) builds the `traceCmd`.
- The last (and trickiest) part is to focus on the `Validate()` changes in `run.go` and how they maintain backwards compatibility.

  Design choices to review
  - The `--selector` flag can only be used with `--tracer`. This is to ensure that users don't mix old and new functionality.
  - The `--program`, `--eval` and `--file` flags are interchangeable and they just populate the `o.program` field eventually (as in the previous implementation.)
  - When using `--selector` all filters like `node`, `pod` and `container` must be passed in via the flag and cannot be provided via positional arguments.
  - When the run command is finding the right entity, it will update the `parsedSelector` with updated `node`, `pod` and `container` values. This removes the local `podUID` field and changes the behaviour of the `container` field to only take flag input and no updates.
  - When using `--selector` the resource filtering is still done through `resourceArg` and the arg value is populated in order of specificity (node -> pod.)
- **Note:** I have strayed from adding tests just yet so that we can align on code changes but will add integration tests once we have initial review.

As an example of how convenient it would be to add a new backend I have reworked #117 on top of this in my own branch https://github.com/zqureshi/kubectl-trace/pull/1 @adelbertc.

@dalehamel @fntlnz @leodido

dalehamel commented on Makefile:
```diff
@@ -9,7 +9,7 @@ GIT_BRANCH ?= $(shell git rev-parse --abbrev-ref HEAD 2>/dev/null)
 
 IMAGE_NAME_INIT ?= quay.io/iovisor/kubectl-trace-init
-IMAGE_NAME ?= quay.io/iovisor/kubectl-trace-bpftrace
+IMAGE_NAME ?= quay.io/iovisor/kubetl-trace-runner
```
```suggestion
IMAGE_NAME ?= quay.io/iovisor/kubectl-trace-runner
```

dalehamel commented on pkg/cmd/run.go:
```diff
@@ -22,7 +22,7 @@ import (
 var (
 	// ImageNameTag represents the default tracerunner image
-	ImageNameTag = "quay.io/iovisor/kubectl-trace-bpftrace:latest"
+	ImageNameTag = "quay.io/iovisor/kubetl-trace-runner:latest"
```
```suggestion
	ImageNameTag = "quay.io/iovisor/kubectl-trace-runner:latest"
```

dalehamel commented on pkg/cmd/run.go:
```diff
@@ -128,10 +139,18 @@ func NewRunCommand(factory factory.Factory, streams genericclioptions.IOStreams)
+	cmd.Flags().StringVarP(&o.filename, "filename", "f", o.filename, "File containing a bpftrace program")
+
+	// flags for new generic interface
+	cmd.Flags().StringVar(&o.tracer, "tracer", "bpftrace", "Tracing system to use")
```
i'm not super familiar with cobra, is it possible to have the list of supported tracers here also? For what users can select from?

zqureshi replied:
I didn't update the documentation yet because there is no other tracer. Do you think it's more appropriate to do that when `bcc` support is added in follow up PR?

zqureshi replied:
I can modify the list here to just include `bpftrace` for now though. Leo had a comment around automatic validation of these things with `go-validator` so will look into that too.

dalehamel commented on pkg/cmd/tracerunner.go:
```diff
@@ -13,15 +13,51 @@ import (
+)
+
+var (
+	bpfTraceBinaryPath = "/bin/bpftrace"
```
is this where we install it to in the container? I think /usr/bin/bpftrace is more common, so we should standardize if that is the case.

zqureshi replied:
No you are correct it is actually installed at `/usr/bin/bpftrace` I copied it without checking from the flag below.

dalehamel commented on pkg/cmd/tracerunner.go:
```diff
@@ -13,15 +13,51 @@ import (
+	//   Select a node by name
+	// - pod
+	//   Select a pod by name
+	// - pod-uid
```
we already handle namespaces elsewhere right?

zqureshi replied:
Yes, namespaces are handled globally by the `kubectl` command.

zqureshi reviewed (commented):
Thanks Dale, I addressed most of your initial feedback.

adelbertc commented on pkg/cmd/run.go:
```diff
@@ -144,30 +163,70 @@ func NewRunCommand(factory factory.Factory, streams genericclioptions.IOStreams)
-		o.container = args[1]
-		if containerFlagDefined {
-			return fmt.Errorf(containerAsArgOrFlagErrString)
+	if o.tracerDefined && containerFlagDefined {
```
I'm a bit lost as to what is happening here, this triggers when we do something like `--tracer=bpftrace` with `-c CONTAINER` ?

zqureshi replied:
Yep, this is to make sure old syntax `-c` and positional args are not mixed with new syntax `--tracer=... --selector=...`

Still a bit apprehensive about keeping both and part of me thinks that once we cut a new release we could maybe just do a breaking change and move over to new syntax. That would make it easier for user to understand semantics long term. WDYT?

adelbertc commented on pkg/cmd/run.go:
```diff
@@ -144,30 +163,70 @@ func NewRunCommand(factory factory.Factory, streams genericclioptions.IOStreams)
+		return fmt.Errorf(resourceAsSelectorErrString)
+	}
+
+	if !o.tracerDefined {
```
One thing that confused me when first using/modifying kubectl-trace was the order in which it parsed things, e.g. when it would magically interpret positional arguments as Pods vs. containers vs. general arguments - would you mind adding some docs here?

zqureshi replied:
Thanks, I can most certainly do that. Based on other review feedback if we decide to keep old flags then I'll add a comment here otherwise if we do a breaking change then I'll rewrite the full flag documentation anyways.

adelbertc commented on pkg/cmd/tracerunner.go:
```diff
@@ -13,15 +13,51 @@ import (
+
+	// Selector (label query) that identifies entity to be traced.
+	// selector = label '=' value [',' labelN '=' valueN ...]
+	// Currently supported labels:
```
Is it possible to put these currently supported values as part of the command line parsing lib so that if someone messes up they get a nice error message?

zqureshi replied:
Will do, see my previous comment regarding breaking change.

adelbertc commented on pkg/cmd/tracerunner.go:
```diff
@@ -116,13 +164,54 @@ func (o *TraceRunnerOptions) Run() error {
+			return nil, nil, err
+		}
+		programPath = path.Join(os.TempDir(), "program-container.bt")
+		r := strings.Replace(string(f), "$container_pid", *pid, -1)
```
This `$container_pid` interpolation is also helpful in BCC programs, could we factor out the string substitution for that into a generic function we can easily re-use in the BCC PR?

zqureshi replied:
Will be working on it and will update my example BCC PR as part of it.

adelbertc reviewed (approved):
Wooo LGTM! Just some nits :-)

zqureshi commented:
I rebased the PR after recent changes and updated the testing to use `go-check` instead of `testify`. I do think that testify provides a more fluid interface and integrates better with editors (at least vscode and vim.)

seanmalloy commented on pkg/cmd/run.go:
```diff
@@ -128,10 +139,18 @@ func NewRunCommand(factory cmdutil.Factory, streams genericclioptions.IOStreams)
 		},
 	}
 
+	// flags for existing usage
 	cmd.Flags().StringVarP(&o.container, "container", "c", o.container, "Specify the container")
-	cmd.Flags().BoolVarP(&o.attach, "attach", "a", o.attach, "Whether or not to attach to the trace program once it is created")
 	cmd.Flags().StringVarP(&o.eval, "eval", "e", o.eval, "Literal string to be evaluated as a bpftrace program")
-	cmd.Flags().StringVarP(&o.program, "filename", "f", o.program, "File containing a bpftrace program")
+	cmd.Flags().StringVarP(&o.filename, "filename", "f", o.filename, "File containing a bpftrace program")
```
If i'm reading this right these  flags will be removed in the future?

If yes then ...

I'd recommend at a minimum marking them as deprecated.

zqureshi replied:
Not decided yet, but if we do then I will mark them all as deprecated.

zqureshi replied:
After looking at future use cases this one should stay alongside `--program`.

seanmalloy commented on pkg/cmd/run.go:
```diff
@@ -128,10 +139,18 @@ func NewRunCommand(factory cmdutil.Factory, streams genericclioptions.IOStreams)
 	}
 
+	// flags for existing usage
 	cmd.Flags().StringVarP(&o.container, "container", "c", o.container, "Specify the container")
```
```suggestion
	cmd.Flags().StringVarP(&o.container, "container", "c", o.container, "DEPRECATED: Specify the container")
```

zqureshi replied:
Thanks, done.

seanmalloy commented on pkg/cmd/run.go:
```diff
@@ -128,10 +139,18 @@ func NewRunCommand(factory cmdutil.Factory, streams genericclioptions.IOStreams)
+	// flags for existing usage
 	cmd.Flags().StringVarP(&o.container, "container", "c", o.container, "Specify the container")
-	cmd.Flags().BoolVarP(&o.attach, "attach", "a", o.attach, "Whether or not to attach to the trace program once it is created")
 	cmd.Flags().StringVarP(&o.eval, "eval", "e", o.eval, "Literal string to be evaluated as a bpftrace program")
```
```suggestion
	cmd.Flags().StringVarP(&o.eval, "eval", "e", o.eval, "DEPRECATED: Literal string to be evaluated as a bpftrace program")
```

zqureshi replied:
Thanks Sean, marked it as deprecated.

seanmalloy commented on pkg/tracejob/selector.go:
```diff
@@ -0,0 +1,77 @@
+
+// Selector represents a label query to select the entity to be traced.
+// Most commonly this entity would resolve into a host, pod or container.
+type Selector struct {
```
I believe there should be a k8s Go package that already defines a label selector. Do you think that should be used instead?

I don't have the link to the appropriate Go package right now.

zqureshi replied:
Thanks Sean, Dale and I did initially think about it but we also want to support more operators than are in the k8s spec so not sure if that would work.

seanmalloy commented on pkg/tracejob/selector.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+// Selector represents a label query to select the entity to be traced.
+// Most commonly this entity would resolve into a host, pod or container.
```
```suggestion
// Most commonly this entity would resolve into a node, pod or container.
```

zqureshi reviewed (commented):
Thanks for the review @seanmalloy I will update the documentation wholesale once I have feedback on the implementation approach.

zqureshi commented:
@fntlnz I have updated the PR and also addressed some feedback from Sean around marking old flags as deprecated.

dalehamel commented:
The work on our private branch diverged a lot from this, so I've made another PR to cut this functionality out, so this is now obsoleted by https://github.com/iovisor/kubectl-trace/pull/161

Note that we have dropped the usage of selectors as defined here, as we found that it didn't deliver the value we had hoped for. A separate notion of a "process selector" however does still exist, and we'll open a separate PR for that.
